### DISCUSSION TITLE
Cannot get joinStores to work correctly

### DISCUSSION THREAD

**Discussion by bauerbua** [state-adapt/state-adapt#47] [Q&A] [answered]
Maybe I am missing the obvious, but this should work right? (The variable names are just for the example, but this is the structure I need)

**Combined store** 
```typescript
export const bananas = adapt(['bananaStore', initialBananas], bananaAdapter)
export const apples = adapt(['appleStore', initialApples], appleAdapter)
export const smoothie = joinStores({ bananas, apples })({
...additional stuff
})()
```

**Consuming component**
```typescript
export class SmoothieComponent {
...
  public bananaStore = bananas
  public appleStore = apples
  public smoothieStore = smoothie
...
}
```

It complies fine but I get _"Error: NG0203: inject() must be called from an injection context such as a constructor, a factory function, a field initializer, or a function used with `EnvironmentInjector#runInContext`."_

I wanted to seperate all stores into their own classes and then inject them into the consuming component, but the I fail to use the joinStores method accordingly, since I cannot use this inside of the joinStores method.
```typescript
private smoothieStore = joinStores({ this.bananaStore, this.appleStore })()
```

So my goal is to combine these stores for additional selectors but also use them on their own. What am I doing wrong here?

**Comment by mfp22** [comment, marked as answer]
```typescript 
bananaStore = adapt(['bananas', initialBananas], bananaAdapter);
appleStore = adapt(['apples', initialApples], appleAdapter)

smoothieStore = joinStores({
  bananas: this.bananaStore,
  apples: this.appleStore,
})() 
```

The error is because `adapt` has to be called inside a component or service to have access to dependency injection.

**Reply by bauerbua** [reply]
Thank you for your fast reply! Yup that worked, but I am now facing the problem, that the type inheritance seems to be overwhelmed. The smoothieStore now sometimes has the following error: _Type instantiation is excessively deep and possibly infinite._

And when I use the smoothieStore in the template

```html
<ng-container *ngIf="smoothieStore.allSmoothies$ | async as allSmoothies">
...
</ng-container>
````

The allSmoothies variable resolves as unknown instead of the actual type. Which leads to compile errors again.

Edit: Webstorm inherits the type inside of the class correctly but does not transfer that accordingly to the template. VSCode complains about both. I already restarted the TSServer... Any ideas?

**Reply by mfp22** [reply]
That doesn't normally happen except in extremely complex features. Make sure you're using TypeScript `strict: true` in tsconfig. If that doesn't work, try doing an object spread somewhere. Like
```typescript 
smoothieStore = {
  ...joinStores({
    bananas: this.bananaStore,
    apples: this.appleStore,
  })(),
};
```

You might also be self-referring something and creating an inference loop.  

If all else fails, you may need to explicity type some things.

But like I said, this almost never happens. The only time I've seen it is when I defined selectors like 10 levels deep.

**Reply by mfp22** [reply]
Oh and even when I did see it I fixed the types and was able to define selectors 20 levels deep.

Maybe typescript behaves differently on different machines. If it takes too long to calculate the inferred types, it will quit and throw that error.

**Reply by bauerbua** [reply]
Thank you for your help. I will keep you posted

**Reply by bauerbua** [reply]
I had different types defined for the state model and the initial state. It compiles now but I am not entirely sure if this was what it fixed it... I also noticed that I had `EntityState<Banana, string>` still defined, which is unnecessary I guess.

WebStorm still does not show the correct type inside of the template but maybe this is just a WebStorm issue, since it compiles just fine now.


EDIT: Ah well it didn't, error is back

**Reply by mfp22** [reply]
Are you using 1.2.1?
The 2nd generic parameter to EntityState should be the property name. Like 'id'

**Reply by bauerbua** [reply]
One shady setup I have found is the following. Maybe this causes the issue.

```typescript
export type Apples = (Apple | PinkLadyApple | GoldenApple);

export interface ApplesStateModel {
  apples: EntityState<Apples>;
}

const appleAdapter = createAdapter<Apples>()({
  ...defaultAppleAdapter,
  ...pinkLadyAppleAdapter,
  ...goldenAppleAdapter,
})

const applesEntityAdapter = createEntityAdapter<Apples>()(appleAdapter);

export const appleAdapter = joinAdapters<ApplesStateModel>()({
  apples: applesEntityAdapter,
})

````

The defaultAppleAdapter and the others currently would look something like this:
```typescript
export const defaultAppleAdapter = createAdapter<Apple>()({
  setColor: (apple, color: string) => ({ ...apple, color }),
  setSize: (apple, size: number) => ({...apple, size}),
});

export const goldenAppleAdapter = createAdapter<GoldenApple>()({
  setCarat: (apple, carat: number) => ({ ...apple, carat }),
});
```

Combining the adapters with just spread operators seems kinda wrong, but I would like to create an adapter, which can handle different types of apples. The different apples also hold different types of properties. E.g. the goldenApple would additionally have the property 'carat', which the other apples would not have. So in short: How can I combine these adapters while still maintaining correct typing?

Thank you for your time and help in advance!

**Reply by bauerbua** [reply]
Well it still does not change anything if I just paste all the methods into one single adapter. I think I might need to rethink my structure to keep it simpler, but I really don't know how currently

**Reply by mfp22** [reply]
The types got more complex after I fixed the last bug, so maybe I need to work on this again. I also think the union type might be adding to the complexity.

Can you create a repro of this issue on StackBlitz? That would help a lot.

**Reply by bauerbua** [reply]
Yes sure. I'll create one. The structure is quite complex indeed. Maybe there is a way for me to make it simpler anyways. I'll post the link as soon as I'm done

**Comment by bauerbua** [comment]
Ok now here is the [Stackblitz](https://stackblitz.com/edit/stackblitz-starters-klfrdl?file=src%2Fapp%2Fsmoothie%2Fsmoothie.component.html)

I noticed while rebuilding my app, that the _Type instantiation is excessively deep and possibly infinite_. error occured as soon as I abstracted the `Fruit` type. So possibly it has something to do with multiple `extends` when building the according interfaces. I hope the structure is kinda clear, otherwise let me know :)

**Reply by bauerbua** [reply]
My acutal implementation of this has some additional actions and selectors, so this is just the base. 

The general requirements are that there are two different fruits which have an overlap in the way they are processed (peeled or unpeeled). Depending on the way they will be processed a user would need to add different properties to the fruit and take different actions in next steps.

**Reply by mfp22** [reply]
Thanks for the info. I'm looking at this now. It's not uncommon in libraries with complex types to run into this issue. But last time I figured out a trick to get it to work, so hopefully I can figure something out this time too in the library itself. https://github.com/microsoft/TypeScript/issues/34933

**Reply by mfp22** [reply]
I think I may have a fix. I copied your StackBlitz code into the StateAdapt repo locally, and got the same error. Then I changed some types and got it to work, as well as some much more complex features. It runs slow, but I need to look into it more. Some of the types are still not exactly what I would like them to be. I will have to look into this further and sort them out. 

If you want to patch this in your own project now, you can open the `join-stores.types.d.ts` file in `node_modules`, find where it is defining `export type StoreBuilder<...> = ...` and replace everything *below* with this code:

```typescript

export type SelectorsWithNewBlock<
    State,
    S extends Selectors<State>,
    NewBlock extends Selectors<SelectorReturnTypes<State, S>>,
  > = {} & ReturnTypeSelectors<State, {} & SelectorReturnTypes<State, S>, {} & S & NewBlock>;

type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19];

export type AddNewBlock<
  State,
  S extends Selectors<State>,
  SE extends StoreEntries,
  D extends Prev[number] = 19,
> = (
  builder: StoreBuilder<State, S, SE>,
) => D extends [never] ? S : NewBlockAdder<State, S, SE, D>;

export interface NewBlockAdder<
  State,
  S extends Selectors<State>,
  SE extends StoreEntries,
  D extends Prev[number] = 19,
> {
  (): {} & JoinedStore<State, S>;

  <NewBlock extends Selectors<SelectorReturnTypes<State, S>>>(
    newBlock: NewBlock,
  ): {} & ReturnType<
    AddNewBlock<
      State,
      {[K in string & keyof SelectorsWithNewBlock<State, S, NewBlock>]: SelectorsWithNewBlock<State, S, NewBlock>[K]},
      SE,
      Prev[D]
    >
  >;
}
```

Messy but it _should_ work. Your actual project may have other issues but this worked for me. Let me know how it goes.

**Reply by bauerbua** [reply]
Great thank you very much. I actually implemented a bit simpler type structure for now in order to avoid this issue. But I will try your code in the next days.

**Reply by mfp22** [reply]
Btw your repro is working with the latest version. I updated it to StateAdapt 2.0 syntax though. https://stackblitz.com/edit/stackblitz-starters-ndmqsb

https://github.com/state-adapt/state-adapt/commit/27981b3b5011dff6d4cccda9a1d187beecf67c2f
